### ISSUE TITLE
Marathon leaks file descriptors

### ISSUE THREAD

**Issue by yangzhares** [mantl/mantl#1460]
**Environment**
**OS**: CentOS Linux release 7.2.1511 (Core)
**Mantl**: 1.0.3

**What's problem**: service 'marathon' check failed by consul
**Error log**: curl: (28) Operation timed out after 5000 milliseconds with 0 out of -1 bytes received
**Check steps**:
- Check marathon 18080 socket connection status, many connections' status is **CLOSE_WAIT**.

``` shell
# netstat -anl | grep 18080 | awk '/^tcp/ {t[$NF]++}END{for(state in t){print state, t[state]} }' 
LISTEN 1
SYN_RECV 6
CLOSE_WAIT 6537
FIN_WAIT1 10
```
- Check how many sockets weren't released

``` shell
# netstat -anpl | grep 18080 | wc -l
6551
```
- Check marathon process's max FD limit

``` shell
# more /proc/31491/limits
......
Max open files 8192 8192 files
......
```
- Check how many FDs were opened by marathon

``` shell
# lsof -p 31491 | wc -l
8266
```

here, we can know current FD is larger than max limit 8192, so consul health check can't create new FD for socket to check marathon's status. although can restart marathon as temporary workaround, don't know the real root cause that made  so many unreleased socket connections. 

so did you guys meet this issue before? how did you fix this issue?

**Comment by yangzhares** [comment]
Marathon log:

```
May 19 03:28:48 mantl-control-02 marathon[31491]: java.io.IOException: Too many open files
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at sun.nio.ch.ServerSocketChannelImpl.accept0(Native Method) ~[na:1.8.0_91]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:422) ~[na:1.8.0_91]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at sun.nio.ch.ServerSocketChannelImpl.accept(ServerSocketChannelImpl.java:250) ~[na:1.8.0_91]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at org.eclipse.jetty.server.ServerConnector.accept(ServerConnector.java:382) ~[marathon:0.13.0]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at org.eclipse.jetty.server.AbstractConnector$Acceptor.run(AbstractConnector.java:551) ~[marathon:0.13.0]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:654) [marathon:0.13.0]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:572) [marathon:0.13.0]
May 19 03:28:48 mantl-control-02 marathon[31491]: #011at java.lang.Thread.run(Thread.java:745) [na:1.8.0_91]
May 19 03:28:48 mantl-control-02 marathon[31491]: [2016-05-19 03:28:48,036] WARN  (org.eclipse.jetty.server.ServerConnector:qtp932118570-30-acceptor-0@54af8e42-ServerConnector@704b3b91{HTTP/1.1,[http/1.1]}{0.0.0.0:18080})
```

**Comment by stevendborrelli** [comment]
Can you do a `ps -ef` to see what all the processes are?

**Comment by yangzhares** [comment]
@stevendborrelli , 31491 is marathon
```
[root@mantl-control-02 log]# ps -ef | grep 31491 | grep -v grep
root     31491     1 10 May13 ?        15:08:44 java -Djava.library.path=/usr/local/lib:/usr/lib:/usr/lib64 -Djava.util.logging.SimpleFormatter.format=%2$s%5$s%6$s%n -Xmx512m -cp /usr/bin/marathon mesosphere.marathon.Main --http_port 18080 --mesos_authentication_principal marathon --zk zk://marathon:8tZqxQhBbl9AUEF6@zookeeper.service.consul:2181/marathon --event_subscriber http_callback --artifact_store file:///etc/marathon/store --mesos_authentication_secret_file /etc/sysconfig/marathon-mesos-authentication-secret --hostname mantl-control-02.node.consul --master zk://mesos:5xtt35yKEHC1oJAF@zookeeper.service.consul:2181/mesos --logging_level info
root     31504 31491 10 May13 ?        15:46:46 logger -p user.info -t marathon[31491]
root     31505 31491  0 May13 ?        00:00:06 logger -p user.notice -t marathon[31491]
```
and i think marathon process is the key point,  it's not necessary to  care other processes. here is the part of the result of command `lsof -i -n -P | grep 31491`, from that, it should be marathon process receives localhost's FIN and replies ACK, but don't send FIN itself.
```
java      31491      root   40u  IPv6  11941106      0t0  TCP 127.0.0.1:18080->127.0.0.1:45011 (CLOSE_WAIT)
java      31491      root   46u  IPv6  11941191      0t0  TCP 127.0.0.1:18080->127.0.0.1:45098 (CLOSE_WAIT)
java      31491      root   47u  IPv6  11941324      0t0  TCP 127.0.0.1:18080->127.0.0.1:45127 (CLOSE_WAIT)
java      31491      root   58u  IPv6  11944811      0t0  TCP 127.0.0.1:18080->127.0.0.1:45519 (CLOSE_WAIT)
java      31491      root   59u  IPv6  11946061      0t0  TCP 127.0.0.1:18080->127.0.0.1:45592 (CLOSE_WAIT)
java      31491      root   61u  IPv6  11948199      0t0  TCP 127.0.0.1:18080->127.0.0.1:45733 (CLOSE_WAIT)
java      31491      root   62u  IPv6  11948060      0t0  TCP 127.0.0.1:18080->127.0.0.1:45630 (CLOSE_WAIT)
java      31491      root   67u  IPv6  11946760      0t0  TCP 127.0.0.1:18080->127.0.0.1:45883 (CLOSE_WAIT)
```

**Comment by stevendborrelli** [comment]
It's probably an issue we need to open with the marathon folks.

Does your marathon health check have a timeout on the curl process?

**Comment by yangzhares** [comment]
here is the "curl: (28) Operation timed out after 5000 milliseconds with 0 out of -1 bytes received", due to no enough FD for marathon process, i can bet every time to get the same error log.

**Comment by Brycelol** [comment]
mesosphere/marathon#4624

**Comment by thomasvincent** [comment]
https://jira.mesosphere.com/browse/MARATHON-1723

**Comment by thomasvincent** [comment]
This appears to be fixed in Marathon 1.3.7

**Comment by thomasvincent** [comment]
Fixed in upgrade to 1.4.2 in branch  upgrade_marathon
